### ISSUE TITLE
Zlib 1.2.12 released

### ISSUE THREAD

**Issue by abcfy2** [zlib-ng/zlib-ng#1217]
Zlib 1.2.12 released. Would you plan to be compatible with this version ?

Thanks.

**Comment by nmoinvaz** [comment]
We already have most of the commits that are in zlib 1.2.12. We are missing exporting some crc32 combine functions AFAKI.

**Comment by mtl1979** [comment]
We already have CI action to test if in compatibility mode we have same exported symbols as stock zlib... That test just needs to be updated to use zlib 1.2.12 as baseline and then fix any differences it reports... Even though we diverged from zlib years ago, we still try to include as many commits as possible and if necessary, adapt them to accommodate any overlapping changes specific to zlib-ng.

We didn't want to change the baseline commit before the final version of zlib 1.2.12 was released as that would have required keeping the test up-to-date with each and every commit that adds or changes any exported symbols.

UPDATE: zlib 1.2.12 fails to build shared library, so it's useless for comparing against zlib-ng

**Comment by mtl1979** [comment]
With zlib 1.2.12.1, I get:

```
+ abidiff --no-added-syms --suppressions test/abi/ignore test/abi/zlib-ec3df00224d4b396e2ac6586ab5d25f673caa4c2-x86_64-linux-gnu.abi btmp2/zlib-built.abi
ELF SONAME changed
Functions changes summary: 0 Removed, 0 Changed (24 filtered out), 0 Added (2 filtered out) functions
Variables changes summary: 3 Removed, 0 Changed, 0 Added (1 filtered out) variables

SONAME changed from '' to 'libz.so.1'

3 Removed variables:

  'const char[71] const deflate_copyright'    {deflate_copyright}
  'const char[50] const inflate_copyright'    {inflate_copyright}
  'char* const[10] const z_errmsg'    {z_errmsg}
```

**Comment by nmoinvaz** [comment]
Hidden in this commit are the changes to crc32 functions exports: https://github.com/madler/zlib/commit/21767c654d31d2dccdde4330529775c6c5fd5389

**Comment by mtl1979** [comment]
I don't think we can support both the old and new versions of the changed crc32 related functions as there is no function overloading in standard C.

**Comment by mtl1979** [comment]
This is what I get locally:

```
+ abidiff --no-added-syms --suppressions test/abi/ignore test/abi/zlib-ec3df00224d4b396e2ac6586ab5d25f673caa4c2-x86_64-linux-gnu.abi btmp2/zlib-built.abi
Functions changes summary: 0 Removed, 2 Changed (53 filtered out), 0 Added (2 filtered out) functions
Variables changes summary: 0 Removed, 0 Changed, 0 Added (1 filtered out) variable

2 functions with some indirect sub-type change:

  [C] 'function uLong crc32_combine_gen64(off64_t)' at crc32_comb.c:98:1 has some indirect sub-type changes:
    return type changed:
      entity changed from 'typedef uLong' to compatible type 'void'
        type name changed from 'unsigned long int' to 'void'
        type size changed from 64 to 0 (in bits)
    parameter 1 of type 'typedef off64_t' changed:
      entity changed from 'typedef off64_t' to 'uint32_t*'
      type size hasn't changed
    parameter 2 of type 'typedef off64_t' was added

  [C] 'function uLong crc32_combine_op(uLong, uLong, uLong)' at crc32_comb.c:108:1 has some indirect sub-type changes:
    parameter 1 of type 'typedef uLong' changed:
      typedef name changed from uLong to uint32_t at stdint-uintn.h:26:1
      underlying type 'unsigned long int' changed:
        entity changed from 'unsigned long int' to compatible type 'typedef __uint32_t' at types.h:42:1
          type name changed from 'unsigned long int' to 'unsigned int'
          type size changed from 64 to 32 (in bits)
    parameter 2 of type 'typedef uLong' changed:
      typedef name changed from uLong to uint32_t at stdint-uintn.h:26:1
      underlying type 'unsigned long int' changed:
        entity changed from 'unsigned long int' to compatible type 'typedef __uint32_t' at types.h:42:1
          type name changed from 'unsigned long int' to 'unsigned int'
          type size changed from 64 to 32 (in bits)
    parameter 3 of type 'typedef uLong' changed:
      entity changed from 'typedef uLong' to 'const uint32_t*'
      type size hasn't changed
```

**Comment by KungFuJesus** [comment]
Hmm, the CRC32 speedups seem like they'd bode well for CPUs pre-carryless multiply or hardware CRC instructions.  The combine function no longer needing a len parameter is interesting, does this mean we'd need to target multiple versions of zlib API for compatibility via a compile time determination?

How long before software / linux distributions actually switch over to this new API?

**Comment by mtl1979** [comment]
Just in the last 24 hours I've read about multiple issues in 1.2.12 (release) and 1.2.12.1 (develop)... I'm not pretty confident that any distribution or major third-party project will use them until the issues are resolved.

**Comment by KungFuJesus** [comment]
lol, except Arch: https://archlinux.org/packages/core/x86_64/zlib/

I kind of wonder if maybe this is why I was getting segfaults with ld when I had accidentally copied the zlib.so.1 symlinks over into /usr/lib64 instead of _just_ the zlib_ng binary.

**Comment by mtl1979** [comment]
> lol, except Arch: https://archlinux.org/packages/core/x86_64/zlib/
> 
> I kind of wonder if maybe this is why I was getting segfaults with ld when I had accidentally copied the zlib.so.1 symlinks over into /usr/lib64 instead of _just_ the zlib_ng binary.

There's always exceptions... I've seen other "bad" packages even in Ubuntu. It all comes to not testing packages properly and thoroughly. If there weren't any issues in 1.2.12, they wouldn't have started 1.2.12.1 so fast... Package managers can also add individual commits from 1.2.12.1 and still call it 1.2.12 until the next official release.

So far, the issues I know are:
1. shared build gives linker errors due to missing -fPIC, soname and version map
2. cross-building doesn't detect that operating system is Linux, so wrong parameters are passed to linker
3. New crc32 functions sometimes return different value compared to old
4. cmake build can fail due to missing/removed files, or deprecated but not removed options

**Comment by nmoinvaz** [comment]
It looks like the crc32 changes add another large table. I haven't checked to see how big those are.

**Comment by mtl1979** [comment]
@nmoinvaz byte-wise about double the size of the "old" tables, so not "too" big...

**Comment by KungFuJesus** [comment]
> * New crc32 functions sometimes return different value compared to old

Ooof, that sounds like a pretty brutal one.  It'd be cool to have an API compatible version of zlib 1.2.12 free of zlib 1.2.12's bugs, though.

**Comment by mtl1979** [comment]
@KungFuJesus It might be possible... I know one of the reasons for difference in crc32 output was not masking out higher 32 bits when long is 64 bits... Our code only use 32-bit integer type, so we didn't/won't introduce that issue...

**Comment by mtl1979** [comment]
The braid tables in the new code are quite large though... 4\*8\*256 or 8\*8\*256 bytes...

**Comment by KungFuJesus** [comment]
It's funny, I thought we were at the end of the memory / processing trade-offs.  That convention died a pretty brutal death in a lot of the floating point math world (except for fixed point embedded systems or FPGAs where lookup tables made a lot of sense).  I imagine, given the numbers boasted for the improvements, it prevents a lot of Galois Field multiplications.

**Comment by mtl1979** [comment]
I'm not quite sure how much we have improved our non-vectorized crc32 implementation since we forked from zlib, but just avoiding excess zeroing of upper bits saves a lot. It also made sense to switch to fixed-size integer types when we focused on performance with 64-bit processors.

**Comment by KungFuJesus** [comment]
Hmm, I was a little bit surprised how painful this was to get compiled (mostly name conflicts and the fact that we rely on gtest existing) but here's a comparison:

### Zlib 1.2.12
```crc32_test/byfour/1             3.95 ns         3.95 ns    177348998
crc32_test/byfour/8             20.3 ns         20.3 ns     34416676
crc32_test/byfour/64             145 ns          145 ns      4818407
crc32_test/byfour/512            290 ns          290 ns      2417090
crc32_test/byfour/4096          1281 ns         1281 ns       546489
crc32_test/byfour/32768         9459 ns         9458 ns        73990
crc32_test/byfour/262144       74821 ns        74820 ns         9355
crc32_test/byfour/2097152     598160 ns       598154 ns         1170
crc32_test/byfour/4194304    1195945 ns      1195939 ns          585
```

### Zlib-ng (develop)
```
crc32/byfour/1                3.78 ns         3.78 ns    185278114
crc32/byfour/8                7.73 ns         7.73 ns     90575127
crc32/byfour/64               55.3 ns         55.3 ns     12651520
crc32/byfour/512               436 ns          436 ns      1605380
crc32/byfour/4096             3482 ns         3482 ns       201013
crc32/byfour/32768           27858 ns        27858 ns        25127
crc32/byfour/262144         222922 ns       222921 ns         3141
crc32/byfour/2097152       1785061 ns      1785051 ns          392
crc32/byfour/4194304       3569303 ns      3569284 ns          196
crc32/pclmulqdq/1             4.67 ns         4.67 ns    151073967
crc32/pclmulqdq/8             7.69 ns         7.69 ns     91044270
crc32/pclmulqdq/64            24.0 ns         24.0 ns     29179267
crc32/pclmulqdq/512           45.3 ns         45.3 ns     15444638
crc32/pclmulqdq/4096           218 ns          218 ns      3205613
crc32/pclmulqdq/32768         1586 ns         1586 ns       441395
crc32/pclmulqdq/262144       12481 ns        12481 ns        56079
crc32/pclmulqdq/2097152     118312 ns       118312 ns         5857
crc32/pclmulqdq/4194304     239686 ns       239685 ns         2863
```

So, it's definitely ~3x faster.  Obviously the pclmulqdq version still beats the pants off of it.

**Comment by mtl1979** [comment]
Our "unoptimized" code is still faster with lengths up to 64 bytes ;)

**Comment by KungFuJesus** [comment]
That is a curious case, I do wonder if that would be true of 1.2.11's implementation as well.  If so, it's likely very little to do with the new set of tables and lookups.

**Comment by mtl1979** [comment]
Gotta benchmark against 1.2.11 then, I guess ;)

**Comment by KungFuJesus** [comment]
lol, I just blew away those changes to make that compile with stock zlib.  From what I can tell from the perf benchmark, their version unrolls an ungodly amount whereas ours is very short.  Given that gist of the optimization is that they are trying to take advantage of the super scalar nature of most architectures, _some_ of that might actually be on purpose.  I can certainly see where it might be faster for shorter string lengths to have loops instead.

**Comment by mtl1979** [comment]
If we want to support both 1.2.11 and 1.2.12 APIs, we need to have both versions anyways... To get best performance, we need to use hybrid version for non-compatibility mode ;)

**Comment by KungFuJesus** [comment]
Well, technically speaking I think you can implement the old interfaces with the new tables, I think you just simply ignore the length parameter for the combine function.  The other exposed functions are otherwise the same.  But I am curious how vanilla 1.2.11's crc fares against <= 64 bytes.

**Comment by mtl1979** [comment]
I'm more interested in how to make compiler accept two C functions with same basename but different signature and version stamp... In a portable way, of course...

**Comment by KungFuJesus** [comment]
Oh that's definitely not possible without some mangling, and it's certainly not portable.  It would have to be a compile time thing that we pretend to be one or the other.  

Of course, if we simply don't use the third parameter, we could maybe take advantage of the fact that the cdecl convention just requires arguments be setup in registers before call.  That seems like a weird abuse of the calling convention, though.  

On the not portable side of things - doesn't glibc maintain multiple function versions?  I'm not sure if the signature on said functions ever change, but I do know they keep around old versions for compatibility going forward.

**Comment by mtl1979** [comment]
I know Linux can do it with gcc and inline assembly, but not sure about clang or VC++...

https://gcc.gnu.org/wiki/SymbolVersioning

**Comment by KungFuJesus** [comment]
Oh here we go: 
https://developers.redhat.com/blog/2019/08/01/how-the-gnu-c-library-handles-backward-compatibility#change_your_abi_with_compatibility

**Comment by KungFuJesus** [comment]
Ahh, the bigger caveat:

```
This code sets the version for all of your symbols that don't already have a specific version to v1. Of course, you only want to do this before your first release, because versioning your symbols is itself an ABI change. Then, as you develop future versions of your library, you add more version clauses and list new symbols in those new clauses.
```

So, if zlib never versioned the symbol in the first place, it's no good to us.

**Comment by mtl1979** [comment]
Not versioning a symbol is not an issue as you can specify empty string as version...

**Comment by KungFuJesus** [comment]
Right, but then doesn't that cause an issue linking to the new version if zlib doesn't decorate the new version with a name?

**Comment by mtl1979** [comment]
When you specify empty version, the linker creates weak symbol that points the undecorated name to the symbol with just "@" at the end...

**Comment by KungFuJesus** [comment]
Right but if you're supporting both the new and old version, which version ends up being the empty one?  If things linking to zlib 1.2.11 don't know to look at @@v1, you'd want that to be the empty version.  If you're targeting something linking to the newer version, then you'd have the empty one to be @@v2.

This is interesting...https://github.com/madler/zlib/blob/master/crc32.c#L654
They are doing 3 independent runs of CRC's and then combining them later.  I wonder how much better that might be in terms of throughput.

**Comment by mtl1979** [comment]
Usually the oldest version will be the default undecorated one, as the new ones already know about the version map used to build the library, so use the versioned symbol.

**Comment by KungFuJesus** [comment]
Right, but if upstream zlib didn't version the symbol we wouldn't be able to either, to maintain API compliance.  It's a shame, this is perfect application of symbol versioning.  I'm not sure if this is doable on PE binaries, though.

**Comment by mtl1979** [comment]
Of course we can version, as we can use the version map to hide all the ugly details... just put the extra stuff in "local" section ;)

**Comment by KungFuJesus** [comment]
Well sure, all of that is fine until ZLIB_COMPAT is defined, though.  There also would seem to be very little reason to want to maintain the old API if they are recompiling against zlib-ng, anyhow.

**Comment by mtl1979** [comment]
We use different version stamps depending on ZLIB_COMPAT contents, but we can also use different version decorations in .c files as preprocessor knows if we are building with compatibility enabled or disabled, so use new API starting from next available zlib-ng version number...

**Comment by KungFuJesus** [comment]
Sure, the zlib-ng proper API can of course do all of these things.  I really wish Mark Adler had done the right thing here, though.  Now that the API has changed the user is going to have to pay close attention to the version their distro is using / application is linked with.

**Comment by mtl1979** [comment]
Mark Adler has never been the smartest developer... Whole zlib project was on hold for several years until work was resumed...

**Comment by KungFuJesus** [comment]
Ahh man, was about to test their arm changes on manjaro ARM and noticed that they don't even do dynamic runtime detection of ACLE, you have to build it by hand:
```
 * Use ARM machine instructions if available. This will compute the CRC about
 * ten times faster than the braided calculation. This code does not check for
 * the presence of the CRC instruction at run time. __ARM_FEATURE_CRC32 will
 * only be defined if the compilation specifies an ARM processor architecture
 * that has the instructions. For example, compiling with -march=armv8.1-a or
 * -march=armv8-a+crc, or -march=native if the compile machine has the crc32
 * instructions.
```
Still, hopefully interesting comparison upcoming.

**Comment by mtl1979** [comment]
Just proves we are smarter than them ;)

**Comment by KungFuJesus** [comment]
Well that was woefully disappointing:
```
[adam@odroid benchmarks]$ make
[ 87%] Built target zlibstatic
[ 87%] Linking CXX executable benchmark_zlib
[100%] Built target benchmark_zlib
[adam@odroid benchmarks]$ taskset -c 2 ./benchmark_zlib --benchmark_filter=crc32
2022-03-31T22:23:39-04:00
Running ./benchmark_zlib
Run on (6 X 1992 MHz CPU s)
Load Average: 0.04, 1.31, 3.86
--------------------------------------------------------------------
Benchmark                          Time             CPU   Iterations
--------------------------------------------------------------------
crc32_test/byfour/1             27.2 ns         27.2 ns     25471148
crc32_test/byfour/8             31.2 ns         31.2 ns     22448543
crc32_test/byfour/64            81.6 ns         81.6 ns      8572570
crc32_test/byfour/512            508 ns          507 ns      1379562
crc32_test/byfour/4096          3796 ns         3795 ns       184468
crc32_test/byfour/32768        18892 ns        18885 ns        37049
crc32_test/byfour/262144      140861 ns       140808 ns         4970
crc32_test/byfour/2097152    1123430 ns      1122344 ns          623
crc32_test/byfour/4194304    2241435 ns      2238861 ns          313
[adam@odroid benchmarks]$ taskset -c 2 ./benchmark_zlib_ours --benchmark_filter=crc32/acle
2022-03-31T22:24:00-04:00
Running ./benchmark_zlib_ours
Run on (6 X 1992 MHz CPU s)
Load Average: 0.16, 1.25, 3.79
-------------------------------------------------------------
Benchmark                   Time             CPU   Iterations
-------------------------------------------------------------
crc32/acle/1             7.34 ns         7.34 ns     95390483
crc32/acle/8             11.0 ns         11.0 ns     63596496
crc32/acle/64            22.0 ns         22.0 ns     31796767
crc32/acle/512            100 ns          100 ns      6992154
crc32/acle/4096           687 ns          687 ns      1018631
crc32/acle/32768         5436 ns         5434 ns       128763
crc32/acle/262144       43017 ns        43000 ns        16279
crc32/acle/2097152     361958 ns       361617 ns         1935
crc32/acle/4194304     723690 ns       722852 ns          968
```

**Comment by mtl1979** [comment]
I still remember drafting the first ACLE-optimized version for zlib-ng... lol... It was pretty much first code I wrote for ARM...

**Comment by KungFuJesus** [comment]
Even on the Cortex-A53, supposedly the CPU this was tested on, the simpler one word at a time CRC32 is faster...I'd have thought they'd try that before trying to capitalize on more ILP.  

We actually could vectorize the loads and reduce that a bit, extracting from the vector register lanes into scalar GPRs each iteration.  It could also allow for a fast copy + crc, I guess it depends on how slow the "extract from lane" moves are.  We also probably ought to modify our version to start loading from aligned boundaries, it makes a pretty significant difference on even newer aarch64 CPUs.  But man, I was fully expecting their version to be at least twice as fast.

**Comment by mtl1979** [comment]
My original version did incrementally use larger sizes until the pointer was aligned, and used reverse logic for tail bytes... basically using least amount of crc instructions...

**Comment by KungFuJesus** [comment]
Oh my bad, it is:
```
    Z_REGISTER const uint16_t *buf2;
    Z_REGISTER const uint32_t *buf4;

    c = ~crc;
    if (len && ((ptrdiff_t)buf & 1)) {
        c = __crc32b(c, *buf++);
        len--;
    }
```
Not that it would have mattered for these tests, they are aligned allocations.  It also probably wouldn't hurt to remove those unrolled versions since really none of those crc's can overlap, anyway.  UNROLL_MORE I don't think is defined but even if it were at best it'd save you a little bit on the branching and not much else. Was that ever measured to be beneficial?

**Comment by mtl1979** [comment]
I'm not sure if we did benchmarks with UNROLL_MORE enabled... I didn't have access to ARM hardware 24/7 back then...

**Comment by nmoinvaz** [comment]
I did benchmarks with `UNROLL_MORE` enabled in [here](https://github.com/zlib-ng/zlib-ng/issues/307#issuecomment-875265330). It really didn't make any difference.

**Comment by Dead2** [comment]
The `UNROLL_MORE` ifdefs were added because it makes no sense to do that on modern cpus because they understand the code better if we rather let them understand the loop and also have a further lookahead on future instructions.

Older cpus however were not able to understand loops like that, and loop unrolling was useful for them.
Nobody has really defined where the cutoff point for that still being useful is though, and nobody has done any extensive testing to figure it out. If I were to hazard a guess, I'd think at least Core2 and newer should not use it, but IDK on arm/ppc/etc.

**Comment by KungFuJesus** [comment]
So from my experience, the PPC970 and it's older cousin the PPC7450/7400 _do_ prefer extensive unrolling, as they have deep pipelines and apparently don't speculate too far out.  However in that particular function, it probably wouldn't have been very beneficial, as every input to the next instruction requires the previous one to finish.  Not that this compiles on that ISA or anything (this is very arm specific), but I doubt any iteration of arm with ACLE this was helpful for.

**Comment by nmoinvaz** [comment]
Is there a reason why the crc32 braiding stuff in 1.2.12 was never merged at the time from develop?

**Comment by mtl1979** [comment]
> Is there a reason why the crc32 braiding stuff in 1.2.12 was never merged at the time from develop?

1. It takes quite a lot of memory
2. The code needs to be cleaned up before merging in
3. Nobody had spare time to do steps 1 and 2

**Comment by KungFuJesus** [comment]
So I _think_ the changes are in to be able to close this issue, no?

**Comment by nmoinvaz** [comment]
Yea I believe so.

**Comment by nmoinvaz** [comment]
Actually only #1289.

**Comment by nmoinvaz** [comment]
Looks like everything is in. We should be fully compatible with 1.2.12 now.

**Comment by KungFuJesus** [comment]
Should be closed by way of #1289 .
